src/Orleans.Core/Streams/Core/IStreamIdMapper.cs

```diff
@@ -0,0 +1,16 @@
+using Orleans.Metadata;
+using Orleans.Runtime;
+
+namespace Orleans.Streams
+{
+    /// <summary>
+    /// Common interface for component that map a StreamId to a GrainId
+    /// </summary>
+    public interface IStreamIdMapper
+    {
+        /// <summary>
+        /// Get the corresponding GrainId from the StreamId
+        /// </summary>
+        GrainId GetGrainId(GrainBindings grainBindings, StreamId streamId);
```

Comment:
I think ultimately a model which is compositional would be better. Eg, TryGetGrainId, so you can run them in sequence before falling back to a default